pkg/services/accesscontrol/permreg/permreg_test.go

```diff
@@ -0,0 +1,240 @@
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func Test_permissionRegistry_RegisterPluginScope(t *testing.T) {
+	tests := []struct {
+		scope     string
+		wantKind  string
+		wantScope string
+	}{
+		{
+			scope:     "folders:uid:AABBCC",
+			wantKind:  "folders",
+			wantScope: "folders:uid:",
+		},
+		{
+			scope:     "plugins:id:test-app",
+			wantKind:  "plugins",
+			wantScope: "plugins:id:",
+		},
+		{
+			scope:     "resource:uid:res",
+			wantKind:  "resource",
+			wantScope: "resource:uid:",
+		},
+		{
+			scope:     "resource:*",
+			wantKind:  "resource",
+			wantScope: "resource:",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.scope, func(t *testing.T) {
+			pr := newPermissionRegistry()
+			pr.RegisterPluginScope(tt.scope)
+			got, ok := pr.kindScopePrefix[tt.wantKind]
+			require.True(t, ok)
+			require.Equal(t, tt.wantScope, got)
+		})
+	}
+}
+
+func Test_permissionRegistry_RegisterPermission(t *testing.T) {
+	tests := []struct {
+		name          string
+		action        string
+		scope         string
+		wantKind      string
+		wantPrefixSet PrefixSet
+		wantSkip      bool
+	}{
+		{
+			name:          "register folders read",
+			action:        "folders:read",
+			scope:         "folders:*",
+			wantKind:      "folders",
+			wantPrefixSet: PrefixSet{"folders:uid:": true},
+		},
+		{
+			name:          "register app plugin settings read",
+			action:        "test-app.settings:read",
+			wantKind:      "settings",
+			wantPrefixSet: PrefixSet{},
+		},
+		{
+			name:          "register an action on an unknown kind",
+			action:        "unknown:action",
+			scope:         "unknown:uid:*",
+			wantPrefixSet: PrefixSet{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pr := newPermissionRegistry()
+			pr.RegisterPermission(tt.action, tt.scope)
+			got, ok := pr.actionScopePrefixes[tt.action]
+			require.True(t, ok)
+			for k, v := range got {
+				require.Equal(t, v, tt.wantPrefixSet[k])
+			}
+		})
+	}
+}
+
+func Test_permissionRegistry_IsPermissionValid(t *testing.T) {
+	pr := newPermissionRegistry()
+	pr.RegisterPermission("folders:read", "folders:uid:")
+	pr.RegisterPermission("test-app.settings:read", "")
+
+	tests := []struct {
+		name    string
+		action  string
+		scope   string
+		wantErr bool
+	}{
+		{
+			name:    "valid folders read",
+			action:  "folders:read",
+			scope:   "folders:uid:AABBCC",
+			wantErr: false,
+		},
+		{
+			name:    "valid folders read with wildcard",
+			action:  "folders:read",
+			scope:   "folders:uid:*",
+			wantErr: false,
+		},
+		{
+			name:    "valid folders read with kind level wildcard",
+			action:  "folders:read",
+			scope:   "folders:*",
+			wantErr: false,
+		},
+		{
+			name:    "valid folders read with super wildcard",
+			action:  "folders:read",
+			scope:   "*",
+			wantErr: false,
+		},
+		{
+			name:    "invalid folders read with wrong kind",
+			action:  "folders:read",
+			scope:   "unknown:uid:AABBCC",
+			wantErr: true,
```

Comment:
nit: could maybe add another test case with invalid attribute (`folders:id:abc`), and also a test case with no scope passed in where scope is expected.